optuna/samplers/_tpe/sampler.py

```diff
@@ -204,34 +283,62 @@ def infer_relative_search_space(
         if not self._multivariate:
             return {}
 
+        if self._group:
+            assert self._group_decomposed_search_space is not None
+            self._search_space_group = self._group_decomposed_search_space.calculate(study)
+            _search_space: Dict[str, BaseDistribution] = {}
+            for sub_space in self._search_space_group.search_spaces:
+                for name, distribution in sub_space.items():
+                    if not isinstance(distribution, _DISTRIBUTION_CLASSES):
+                        self._log_independent_sampling(study, trial, name)
+                        continue
+                    _search_space[name] = distribution
+            return _search_space
+
         search_space: Dict[str, BaseDistribution] = {}
         for name, distribution in self._search_space.calculate(study).items():
             if not isinstance(distribution, _DISTRIBUTION_CLASSES):
-                if self._warn_independent_sampling:
-                    complete_trials = study.get_trials(deepcopy=False)
-                    if len(complete_trials) >= self._n_startup_trials:
-                        self._log_independent_sampling(trial, name)
+                self._log_independent_sampling(study, trial, name)
                 continue
             search_space[name] = distribution
 
         return search_space
 
-    def _log_independent_sampling(self, trial: FrozenTrial, param_name: str) -> None:
-        _logger.warning(
-            "The parameter '{}' in trial#{} is sampled independently "
-            "instead of being sampled by multivariate TPE sampler. "
-            "(optimization performance may be degraded). "
-            "You can suppress this warning by setting `warn_independent_sampling` "
-            "to `False` in the constructor of `TPESampler`, "
-            "if this independent sampling is intended behavior.".format(param_name, trial.number)
-        )
+    def _log_independent_sampling(self, study: Study, trial: FrozenTrial, param_name: str) -> None:
+        if self._warn_independent_sampling:
+            complete_trials = study.get_trials(deepcopy=False)
+            if len(complete_trials) >= self._n_startup_trials:
+                _logger.warning(
+                    "The parameter '{param_name}' in trial#{trial.number} is sampled "
+                    "independently instead of being sampled by multivariate TPE sampler. "
+                    "(optimization performance may be degraded). "
+                    "You can suppress this warning by setting `warn_independent_sampling` "
+                    "to `False` in the constructor of `TPESampler`, "
+                    "if this independent sampling is intended behavior."
+                )
```

Comment:
I changed the logic of this method to reduce redundancy. This change is reasonable because `log_independent_sampling` is called only when `_warn_independent_sampling` is `True`.

Reply:
Note that you can cache the trial count. It's not that critical since this method isn't called in the normal case.